### ISSUE TITLE
NullPointerException in WithOptimisticLock

### ISSUE THREAD

**Issue by skynetsonar** [factcast/factcast#1884]
**Describe the bug**
It seems there is a bug in WithOptimisticLock class. Change on line 57 is producing null:

StateToken token = store.currentStateFor(factSpecs);

We get NullPointerException on line 76:
if (store.publishIfUnchanged(r.factsToPublish(), Optional.of(token)))

Everything still worked fine in version 0.5.0.

**FactCast Server/Client Version**
0.5.1

**Comment by uweschaefer** [comment]
could you share a demo project?

**Comment by uweschaefer** [comment]
A failing test would be an alternative

**Comment by uweschaefer** [comment]
We changed the logic for creating and validating lock tokens in 0.5.1. This involved adding a new RPC service. This is why using this change on the client when talking to an older server will fail (with an UNIMPLEMENTED Exception).

A trivial test like:

```
@FactcastTestConfig(factcastVersion = "0.5.1")
@SpringBootTest
@EnableAutoConfiguration(exclude = {DataSourceAutoConfiguration.class})
@ContextConfiguration(classes = {Application.class})
@DirtiesContext(classMode = DirtiesContext.ClassMode.BEFORE_EACH_TEST_METHOD)
public class StateTokenTest extends AbstractFactCastIntegrationTest {
  @Autowired FactCast fc;

  @Test
  public void issue1884() throws Exception {
    var fspec = FactSpec.ns("foo").type("bar");

    fc.lock(fspec)
        .optimistic()
        .attempt(() -> Attempt.publish(Fact.builder().ns("foo").type("bar").buildWithoutPayload()));
  }
}
```
seems to work fine. Can you provide some more input @skynetsonar ?

**Comment by skynetsonar** [comment]
Thanks for the response @uweschaefer!

Upon further investigation, it seems that the NullPointerException in our tests can be avoided if we adapt our mocking setup to the changed token logic.

But still in our integration tests, we get the UNIMPLEMENTED exception you describe above when we use FactCast server <= 0.5.0.

Unfortunately, our production server still runs on 0.3.8. This works fine with client version 0.5.0. So we can just stay on the 0.5.0 client for the moment. Will #1887 enable us to upgrade the client again even if the server stays at the current version?

**Comment by uweschaefer** [comment]
Thanks for coming back to us and directing our focus in that area.

Investigating your NPE (that i now realize is a failing test where behavior was changed that your mock cannot know about), we realized the incompatibility issue. This is why #1187 was implemented to provide a fallback behavior and negotiate the use of the stateTokenRetrieval on handshake. 
Also we added this warning to the release notes of 0.5.1.

When 0.5.2 is released, it will be able to talk to older servers again without the need to configure or change anything.
Regression tests have been added as well. 

thx

PS: There have been significant improvements in terms of performance and scalability from 0.3 to 0.5, and also lots of CVEs in dependent libs (like the recent RCE + DoS problems in Spring Boot), so maybe you want to consider upgrading at some point.
